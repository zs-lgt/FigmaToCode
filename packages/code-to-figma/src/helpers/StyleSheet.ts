import { CommonStyle } from '../types';
import { Assign } from 'utility-types';

type StyleProp<P, T> = { [P in keyof T]: Partial<CommonStyle> };

const absoluteFill: Partial<CommonStyle> = {
    position: 'absolute',
    left: 0,
    top: 0,
    width: '100%',
    height: '100%'
};

export class StyleSheet {
    static create<T extends StyleProp<any, T>>(styles: T): T {
        return styles;
    }

    static flatten<C extends Partial<CommonStyle>>(styles: C | C[] | void): C {
        if (Array.isArray(styles)) {
            return styles.reduce((acc, item) => ({ ...acc, ...item }), {} as any);
        } else {
            return styles || ({} as C);
        }
    }

    static resolve<T extends { [key: string]: Partial<CommonStyle> }>(value: T): T {
        return value;
    }

    static compose<C1 extends Partial<CommonStyle>, C2 extends Partial<CommonStyle>>(
        style1: C1,
        style2: C2
    ): Assign<C1, C2> {
        return { ...(style1 || ({} as C1)), ...(style2 || ({} as C2)) };
    }

    static hairlineWidth = 1;

    static absoluteFill = absoluteFill;

    static absoluteFillObject = absoluteFill;
}
