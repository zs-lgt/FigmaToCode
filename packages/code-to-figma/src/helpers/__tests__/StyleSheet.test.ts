import { StyleSheet } from '../StyleSheet';

describe('StyleSheet', () => {
    it('compose', () => {
        const style1 = {
            width: '100%',
            backgroundColor: 'green'
        };

        const style2 = {
            height: '100%',
            backgroundColor: 'red'
        };

        const result = StyleSheet.compose(
            style1,
            style2
        );
        expect(result).toMatchSnapshot();
    });

    it('flatten', () => {
        var styles = StyleSheet.create({
            listItem: {
                flex: 1,
                fontSize: 16,
                color: 'white'
            },
            selectedListItem: {
                color: 'green'
            }
        });

        const result = StyleSheet.flatten([styles.listItem, styles.selectedListItem]);
        expect(result).toMatchSnapshot();
    });
});
