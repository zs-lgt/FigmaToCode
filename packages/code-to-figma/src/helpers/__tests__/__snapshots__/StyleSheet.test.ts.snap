// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet compose 1`] = `
Object {
  "backgroundColor": "red",
  "height": "100%",
  "width": "100%",
}
`;

exports[`StyleSheet flatten 1`] = `
Object {
  "color": "green",
  "flex": 1,
  "fontSize": 16,
}
`;
